server/src/lib/server.rs

```diff
@@ -371,43 +437,132 @@ async fn instance_serial(
+    if context.serial_detach.as_ref().map_or(false, |s| !s.is_closed()) {
+        return Err(HttpError::for_unavail(
+            None,
+            "serial console already attached".to_string(),
+        ));
+    }
 
-    // Backpressure: Wait until either the buffer has filled completely, or
-    // (for partial reads) 10ms have elapsed. This prevents spinning on a
-    // serial console which is emitting single bytes at a time.
-    let _ = tokio::time::timeout(
-        core::time::Duration::from_millis(10),
-        context.serial.read_buffer_full(),
-    )
-    .await;
-
-    // The buffer may or may not actually have any output - we've already
-    // introduced our backpressure mechanism with the previous tokio timeout,
-    // so don't bother blocking on this particular read.
-    let mut output = [0u8; 4096];
-    let n = futures::select! {
-        result = context.serial.read(&mut output).fuse() => {
-            result.map_err(|e| {
-                HttpError::for_internal_error(format!(
-                    "Cannot read from serial: {}",
-                    e
-                ))
-            })?
-        },
-        default => { 0 }
-    };
+    let request = &mut *rqctx.request.lock().await;
 
-    let input = request.into_inner().bytes;
-    if !input.is_empty() {
-        context.serial.write_all(&input).await.map_err(|e| {
-            HttpError::for_internal_error(format!(
-                "Cannot write to serial: {}",
-                e
-            ))
+    if !request
+        .headers()
+        .get(header::CONNECTION)
+        .and_then(|hv| hv.to_str().ok())
+        .map(|hv| {
+            hv.split(|c| c == ',' || c == ' ')
+                .any(|vs| vs.eq_ignore_ascii_case("upgrade"))
+        })
+        .unwrap_or(false)
+    {
+        return Err(HttpError::for_bad_request(
+            None,
+            "expected connection upgrade".to_string(),
+        ));
+    }
+    if !request
+        .headers()
+        .get(header::UPGRADE)
+        .and_then(|v| v.to_str().ok())
+        .map(|v| {
+            v.split(|c| c == ',' || c == ' ')
+                .any(|v| v.eq_ignore_ascii_case("websocket"))
+        })
+        .unwrap_or(false)
+    {
+        return Err(HttpError::for_bad_request(
+            None,
+            "unexpected protocol for upgrade".to_string(),
+        ));
+    }
+    if request
+        .headers()
+        .get(header::SEC_WEBSOCKET_VERSION)
+        .map(|v| v.as_bytes())
+        != Some(b"13")
+    {
+        return Err(HttpError::for_bad_request(
+            None,
+            "missing or invalid websocket version".to_string(),
+        ));
+    }
+    let accept_key = request
+        .headers()
+        .get(header::SEC_WEBSOCKET_KEY)
+        .map(|hv| hv.as_bytes())
+        .map(|key| handshake::derive_accept_key(key))
+        .ok_or_else(|| {
+            HttpError::for_bad_request(
+                None,
+                "missing websocket key".to_string(),
+            )
```

Comment:
Yea, I think we came to the conclusion that it was worth trying outside of dropshot first to even see if that's possible and then maybe lift some common parts to dropshot. But that's going to require a bit more careful design because currently there's a lot of flexibility in how you might want to setup a websocket (or any upgraded) connection which can quickly turn into a lot of complexity on the dropshot side.

For now, I think it's ok to keep this in propolis. We could actually reduce a lot of this boilerplate if we used hyper-tungstenite here except it's a little limited currently. But that should be relatively easy to fix upstream so I think I'll try that.

Reply:
Sounds good to me - let's remember this when we find the second websockets consumer elsewhere at Oxide, then we can/should revisit.